File: 19th/코딩테스트 기초반/week7/8989번_시계.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    // 빠른 입출력
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int T;
    cin >> T;

    string t[5]; // 5개의 시간
    while (T--) {
        vector<pair<int, pair<int, int>>> v; // 각도, 시간, 시간의 idx 저장
        for (int i = 0; i < 5; i++) {
            cin >> t[i];
            int h = stoi(t[i].substr(0, 2)); // 시간 부분
            int m = stoi(t[i].substr(3, 2)); // 분 부분
            int time = h * 60 + m; // 각도가 같을 경우 시간으로 오름차순 해야함

            h %= 12; // 12로 나머지 연산을 하여 12시간 이내로 맞춤

            // 시계를 720등분
            int pos_h = h * 60 + m; // 시침의 위치는 (시간 * 60) + 분
            int pos_m = m * 12; // 분침의 위치는 (분 * 12)

            // 시침과 분침의 차이
            int diff = abs(pos_h - pos_m);
            int angle = min(diff, 720 - diff); // 각도는 180도보다 작아야 함

            v.push_back({ angle, { time, i } }); // 각도, 시간, 시간의 idx
        }

        sort(v.begin(), v.end());

        int ans_pos = v[2].second.second; // 정답 시간의 idx 추출
        cout << t[ans_pos] << "\n";
    }
}

File: 19th/코딩테스트 기초반/week7/8989번_시계.py
for _ in range(int(input())):
    time = list(input().split())
    ans = []

    for i in time:
        h, m = map(int, i.split(':'))  # 시:분 분리

        # 시침 위치를 분 단위로 변환 (12시간 기준)
        pos_h = h % 12
        pos_h = pos_h * 60 + m

        # 분침 위치를 시침 스케일로 변환
        pos_m = m * 12

        # 두 바늘 사이의 각도 계산
        angle = abs(pos_h - pos_m)
        angle = min(angle, 720 - angle)  # 시계 반대쪽 각도 고려
        
        # (각도, 시, 분, 원본 문자열) 튜플로 저장
        ans.append((angle, h, m, i))
    
    # 각도 기준 정렬, 각도가 같으면 시, 분 순으로 정렬
    ans.sort(key=lambda x: (x[0], x[1], x[2]))

    # 세 번째로 작은 각도를 가진 시각 출력
    print(ans[2][3])